Spike sorting v1 pipeline

khl02007 opened the pull request:
Add new sub-package called `v1` within `spikesorting` that contains the following modules:
- `recording.py`
- `artifact.py`
- `sorting.py`
- `curation.py`
- `metric_curation.py`
- `figurl_curation.py`
Also add the following modules in the `spikesorting` package:
- `imported.py`
- `merge.py`
# General
- All the tables live in new schema with naming convention `spikesorting_v1_modulename`
- Make the name of tables singular (`~Parameters` $\to$ `~Parameter`)
- Remove environment variables for saving preprocessed recordings, sortings, and waveforms; these are now saved in the same directory as other analysis NWB files
- Reorder steps in `make` methods to have the following structure:
	- Fetch (fetch info from database) - Do (process fetched info) - Insert (insert info back to database)
- Make most methods `classmethod`
- Add `neuroconv` to installation requirement (needed to write recording to NWB)
# Module-specific
## `recording.py`
### `SortGroup`
- Move `Session` foreign key from primary to secondary key
- Remove methods 
	- `SortGroup.set_group_by_electrode_group`
	- `SortGroup.set_reference_from_list`
	- `SortGroup.get_geometry`
### `SortInterval`
- Remove; use `IntervalList` instead
### `SpikeSortingPreprocessingParameter`
- Change parent class to `dj.Lookup`
- Add default parameters to `contents` property of the table
### `SpikeSortingRecording`
- Use `recording_id` as the sole primary key; move other attributes that used to be primary key to secondary
- Write recording to NWB
- Add `SpikeSortingRecording.get_recording` method
## `artifact.py`
### `ArtifactDetectionParameter`
- Change parent class to `dj.Lookup`
- Add default parameters to `contents` property
### `ArtifactRemovedIntervalList`
- Removed
### `ArtifactInterval`
- Rename to `ArtifactRemovedInterval`
- Add new primary key: `artifact_id`
- No longer computes artifact intervals; only computes intervals free of artifacts
- No longer saves artifact free intervals; these are saved in `IntervalList` and interval_list_name is same as artifact_id
## `sorting.py`
### `SpikeSorterParameter`
- Change parent class to `dj.Lookup`
### `SpikeSortingSelection`
- Inherit from `IntervalList` for artifact-free sorting interval
### `SpikeSorting`
- Save sorting to an analysis NWB file and remove `SPYGLASS_SORTING_DIR`
- Remove clean up methods for `SPYGLASS_SORTING_DIR`
## `curation.py`
### `Curation`
- Change the secondary key to include `parent_curation_id`, `merges_applied`, and foreign key from `AnalysisNwbfile`
- `dj.Manual` table but using `insert_curation` method is recommended to populate the table; this function saves the sorting and curation information in NWB
- Four methods exposed to the user:
	- `insert_curation`
	- `get_recording`
	- `get_sorting`
	- `get_merged_sorting`: differs from `get_sorting` in that it applies merge groups to the sorting
## `metric_curation.py`
- Move the automatic curation pipleline that used to be in `spikesorting_curation.py` to here
- Merge the waveform extraction, metric computation, and curation labels / merge group generation to a single compute table
### `MetricCuration`
- Save the waveforms, quality metrics, curation labels and merge groups to NWB
- Has methods for accessing these from NWB
## `figurl_curation.py`
- Creates new `curation_uri.json` each time. This contains the curation labels and merge groups to be shown in the figURL.
### `FigURLCurationSelection`
- Contains `metrics_figurl` as primary key; specifies which metrics to include in the figURL
- Has `generate_curation_uri` method, which can be used to prepopulate the curation labels or merge groups from previous curation in the new figURL
### `FigURLCuration`
- Only secondary key is `url`
# To do
- Provide a way to access waveforms as `spikeinterface.WaveformExtractor` from `MetricCuration` table
- Provide a way to include similarity matrices to the figURL
- Write tutorial notebooks
- `pynwb.units` requires the unit ID to be `int`; `sortingview` requires unit IDs to be `str`; `spikeinterface ` allows them to be either type. Write tests to make sure these conventions are respected.

khl02007 commented:
Sorry for the huge PR; happy to make any changes. Working on the items in `To do` but I think ready to be tried.

samuelbray32 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,730 @@
+class SortGroup(dj.Manual):
+    definition = """
+    # Set of electrodes to spike sort together
+    sort_group_id: str  # identifier for a group of electrodes
```
Is there a benefit to having this a `str` instead of the `int` it was before?

khl02007 replied:
I thought making it `str` would enable more descriptive names, e.g. `'probe shank 1'` or `'tetrode 1'` seems better than `1`. But I don't have a strong opinion about this and I'm happy to change it back to int.

edeno replied:
I think it'll be easier to just have int in this case to keep consistency.

khl02007 replied:
I thought more about this and I would prefer to make this `varchar`. I think doing so is more consistent because `electrode_group_name` from `ElectrodeGroup` table is also `varchar`. Using a number to refer to an electrode group is not very descriptive.

samuelbray32 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,730 @@
+    # Set of electrodes to spike sort together
+    sort_group_id: str  # identifier for a group of electrodes
+    ---
+    -> Session
```
Sorry if I'm missing something, but with Session as a secondary key, doesn't this mean that every sort_group_id needs to be unique, even across Sessions?  

For example, I can't make a `sort_group_id = 0` corresponding to electrode group id 0 for `Winnie20220713_.nwb` and also  a `sort_group_id = 0` corresponding to electrode group id 0 for `Winnie20220714_.nwb` 

Or is the idea to use the naming differently?

khl02007 replied:
oops, you're right, I think I should have kept the foreign key from Session as primary key. I will make this change.

samuelbray32 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,730 @@
+
+
+@schema
+class SpikeSortingPreprocessingParameter(dj.Lookup):
```
Just a style question, but why have singular for parameter? It looks like most the other pipelines use plural

khl02007 replied:
Are you asking why Parameter tables specifically are singular or why table names in general (in this pipeline) are singular?

I made this decision thinking that the convention recommended by datajoint is to make table names singular. I thought that was in the [documentation](https://datajoint.com/docs/core/datajoint-python/0.14/design/tables/declare/)... but can no longer find it, so I may have hallucinated that! (@CBroz1 was there ever such a convention?) I don't have a strong opinion about this but figured it would be easier if every table name was singular since most table we already have are singular.

CBroz1 replied:
I vaguely recall that too - but it might have been a stylistic preference Dimitri expressed verbally. I'd prioritize alignment with existing tables. Doing a global grep on the repo (`class .*Param.*\(dj`), I'm seeing plural `Parameters`, or more recent `Params`

khl02007 replied:
Switching back to `Parameters`

samuelbray32 replied:
Should the keys also be plural like they were before?

khl02007 replied:
yes switched the key `preproc_param` back to `preproc_params`.

samuelbray32 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,730 @@
+
+
+@schema
+class SpikeSortingRecording(dj.Computed):
```
I should probably wait for the tutorial, but what is the correct way to populate this table? Since `recording_id` is the primary key, but it isn't defined until inside of the make function.  This means that when I try to populate it with the `SpikeSortingRecordingSelection` key I get this error:

> DataJointError: A table must have dependencies from its primary key for auto-populate to work

khl02007 replied:
My thought was that this table would be populated by calling the `make` method (i.e. `SpikeSortingRecording.populate()` which calls `make`) like any other `dj.Computed` table. Did that not work?

samuelbray32 replied:
Not when I try it.  I'm pretty sure that for that to work `SpikeSortingRecordingSelection` needs to be a primary key (which I think is the normal standard). Other wise when you call populate it has nothing to start from.  For example, after populating `SpikeSortingRecordingSelection`, when I run this

`SpikeSortingRecording().populate()`

I get this error

> DataJointError: A table must have dependencies from its primary key for auto-populate to work

khl02007 replied:
Ah OK, my mistake. Will fix this. Maybe I have to introduce `recording_id` at the selection table.

khl02007 replied:
Introduced new `insert_selection` method to generate `recording_id`. I added docstring to the table definition so that users would know to use it. The motivation for doing this is to avoid the problem of having a very large composite primary key downstream. Let me know if this solution seems satisfactory to you.

samuelbray32 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,730 @@
+                "nwb_file_name": key["nwb_file_name"],
+                "interval_list_name": key["sort_interval_name"],
+            }
+        ).fetch1("sort_interval")
```
`sort_interval_name` is no longer a key in the `SpikeSortingRecordingSelection`. I think this key should be `interval_list_name` now.  This applies in the next fetch as well

`sort_interval` isn't a key in IntervalList

khl02007 replied:
Thanks, I will fix this.

khl02007 replied:
Fixed

khl02007 commented on src/spyglass/spikesorting/v1/__init__.py:
```diff
@@ -0,0 +1,8 @@
+    SpikeSortingRecording,
+)
+
+from .sorting import SpikeSorterParameter, SpikeSortingSelection, SpikeSortingV1
```
Thanks for catching that. I will fix it.

khl02007 replied:
Fixed

khl02007 commented on .vscode/settings.json:
```diff
@@ -1,14 +1,12 @@
 {
```
I mainly wanted to remove `"editor.multiCursorModifier": "ctrlCmd"` because I use a different key. Are you OK with removing that line? I don't care much about the other changes.

edeno replied:
That's fine.

khl02007 replied:
Fixed

CBroz1 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+    proportion_above_thresh=1.0,
+):
+    # create a local dict per worker
+    worker_ctx = {}
+    if isinstance(recording, dict):
+        worker_ctx["recording"] = si.load_extractor(recording)
+    else:
+        worker_ctx["recording"] = recording
+    worker_ctx["zscore_thresh"] = zscore_thresh
+    worker_ctx["amplitude_thresh_uV"] = amplitude_thresh_uV
+    worker_ctx["proportion_above_thresh"] = proportion_above_thresh
+    return worker_ctx
```
```suggestion
    return dict(
        recording=si.load_extractor(recording)
        if isinstance(recording, str)
        else recording,
        zscore_thresh=zscore_thresh,
        amplitude_thresh_uV=amplitude_thresh_uV,
        proportion_above_thresh=proportion_above_thresh,
    )
```

CBroz1 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+
+
+def _init_artifact_worker(
+    recording,
```
I see some methods have type hints but not all. Is the pattern intentional?

khl02007 commented on environment.yml:
```diff
@@ -36,4 +36,5 @@ dependencies:
       - pyyaml
       - click
       - "black[jupyter]"
+      - neuroconv
```
OK I will just copy over the class instead of adding neurconv to dependency.

CBroz1 reviewed (commented):
I reviewed `imported`, `merge`, `__init__`, `artifact` and `curation` - I think a lot of my suggestions could be applied across other files. Let me know what you think. We can decide later if I should PR to this branch with similar edits across all files. 

One topic worth addressing is the question of how much redundancy to allow in the codebase. I'd take a 'don't repeat yourself' approach, but we can talk about the pros/cons when it comes to subpackage versioning

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+
+        # INSERT
+        # - into IntervalList
+        tmp_key = {}
+        tmp_key["nwb_file_name"] = key["nwb_file_name"]
+        tmp_key["interval_list_name"] = key["artifact_id"]
+        tmp_key["valid_times"] = artifact_removed_valid_times
+        IntervalList.insert1(tmp_key, skip_duplicates=True)
```
Agreed, thanks, I have committed this suggestion.

CBroz1 commented on src/spyglass/spikesorting/v1/sorting.py:
```diff
@@ -0,0 +1,318 @@
+
+    @classmethod
+    def insert_default(cls):
+        cls.insert1(cls.contents, skip_duplicates=True)
```
I don't think this will ever actually insert. Lookup gets the contents as soon as it's declared. If we want to keep based on assumptions people will have from other similarly structured tables, I would...
```suggestion
        pass
```

khl02007 replied:
I kept this in case someone deletes the row from `contents` and it needs to be re-inserted. Is this not possible?

edeno commented:
How does this PR affect current users of spikesorting?

khl02007 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,730 @@
+from typing import Tuple, Iterable, Optional, Union, List
+import uuid
```
removed

khl02007 commented on src/spyglass/spikesorting/v1/curation.py:
```diff
@@ -0,0 +1,394 @@
+
+import datajoint as dj
+import pynwb
+import numpy as np
```
removed

khl02007 commented on src/spyglass/spikesorting/v1/sorting.py:
```diff
@@ -0,0 +1,318 @@
+from spyglass.common.common_lab import LabMember, LabTeam
+from spyglass.common.common_interval import IntervalList
+from spyglass.common.common_nwbfile import AnalysisNwbfile
+from spyglass.utils.dj_helper_fn import fetch_nwb
```
Removed

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+        recording = se.read_nwb_recording(
+            recording_analysis_nwb_file_abs_path, load_time_vector=True
+        )
+        if not artifact_param["job_kwargs"]:
```
applied

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+    verbose: bool = False,
+    **job_kwargs,
+):
+    """Detects times during which artifacts do and do not occur.
+    Artifacts are defined as periods where the absolute value of the recording signal exceeds one
+    or both specified amplitude or zscore thresholds on the proportion of channels specified,
+    with the period extended by the removal_window_ms/2 on each side. Z-score and amplitude
+    threshold values of None are ignored.
```
Thanks, applied

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+    IntervalList,
+    _union_concat,
+    interval_from_inds,
+    interval_set_difference_inds,
```
Removed

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+        # FETCH:
+        # - artifact parameters
+        # - recording analysis nwb file
+        artifact_param = (ArtifactDetectionParameter & key).fetch1(
+            "artifact_param"
+        )
+        recording_analysis_nwb_file = (SpikeSortingRecording & key).fetch1(
+            "analysis_nwb_file"
+        )
```
suggestion applied

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+    valid_timestamps = recording.get_times()
+
+    # if both thresholds are None, we skip artifract detection
+    if (amplitude_thresh_uV is None) and (zscore_thresh is None):
```
suggestion applied

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+
+    # if both thresholds are None, we skip artifract detection
+    if (amplitude_thresh_uV is None) and (zscore_thresh is None):
+        recording_interval = np.asarray(
+            [valid_timestamps[0], valid_timestamps[-1]]
+        )
+        artifact_times_empty = np.asarray([])
+        print(
+            "Amplitude and zscore thresholds are both None, skipping artifact detection"
+        )
+        return recording_interval, artifact_times_empty
```
suggestion applied

khl02007 commented on src/spyglass/spikesorting/v1/artifact.py:
```diff
@@ -0,0 +1,485 @@
+    n_jobs = ensure_n_jobs(recording, n_jobs=job_kwargs.get("n_jobs", 1))
+    print(f"Using {n_jobs} jobs...")
+
+    func = _compute_artifact_chunk
+    init_func = _init_artifact_worker
+    if n_jobs == 1:
+        init_args = (
+            recording,
+            zscore_thresh,
+            amplitude_thresh_uV,
+            proportion_above_thresh,
+        )
+    else:
+        init_args = (
+            recording.to_dict(),
+            zscore_thresh,
+            amplitude_thresh_uV,
+            proportion_above_thresh,
+        )
+
+    executor = ChunkRecordingExecutor(
+        recording,
+        func,
+        init_func,
+        init_args,
+        verbose=verbose,
+        handle_returns=True,
+        job_name="detect_artifact_frames",
+        **job_kwargs,
+    )
```
suggestion applied

khl02007 commented on src/spyglass/utils/misc.py:
done

khl02007 commented on src/spyglass/utils/misc.py:
```diff
@@ -0,0 +1,18 @@
+        how many digits of uuid4 to keep
+    """
+    uuid4 = uuid.uuid4()
+    nwb_uuid = nwb_file_name + "_" + initial + "_" + uuid4[:len_uuid]
+    return nwb_uuid
```
I agree it's not ideal. The goal was to use unique IDs as the sole primary key for some of the tables (e.g. `recording_id` for `SpikeSortingRecording`, `sorting_id` for `SpikeSorting`) so that we don't run into the problem of having very long composite primary key downstream. I thought putting the raw nwb file name and the initial in front of the partial uuid would give one some sense of where the data came from and what type of analysis nwb file this is while making sure it is unique. If you have a better approach, I would be happy to hear it.

samuelbray32 replied:
Would the nwb_file_name, interval_list_name, etc. still be in secondary keys? If not it could be difficult to fetch the spikesorting data you need during analysis

khl02007 replied:
@samuelbray32 
Right now, in the case of `SpikeSortingRecording`, its primary key is `recording_id` (inherited from `SpikeSortingRecordingSelection`). Its secondary keys are `analysis_file_name` and `object_id`, corresponding to the output of the `populate` call. The other critical information (`nwb_file_name`, `interval_list_name` and so on) is stored in `SpikeSortingRecordingSelection` as secondary key. so you would fetch it from there e.g. `(SpikeSortingRecordingSelection & [recording_id]).fetch1('nwb_file_name')`.

I initially had these as primary key in  `SpikeSortingRecordingSelection` (as in v0 pipeline) and secondary (foreign) key in `SpikeSortingRecording` as you suggest here, but then I lose the primary key dependency between `SpikeSortingRecordingSelection` and `SpikeSortingRecording` (as you have pointed out in a different comment) and `populate` doesn't work. So my workaround has been to introduce the `insert_selection` method (that I described in yet another comment thread). 

Does this seem OK to you?

samuelbray32 commented on src/spyglass/spikesorting/v1/sorting.py:
```diff
@@ -0,0 +1,318 @@
+
+
+@schema
+class SpikeSorterParameter(dj.Lookup):
```
Similar to other conversation, do we want the table name and keys to stay plural for consistency?

khl02007 replied:
Yep will change these to plural as well

khl02007 replied:
done

edeno commented:
Also another note, I think we should make sure the primary keys have the correct varchar length in this pipeline.

khl02007 commented:
@lfrank @edeno @CBroz1 I have confirmed that the pipeline works (tried with one of Alison's NWB files) and attached a barebones tutorial notebook. Doesn't mean it is complete (I still need to address some of the comments) but everything else can be changed later without affecting the user much (I think). 

@samuelbray32 could you please try the tutorial notebook and let me know if you run into problems?

CBroz1 commented:
Thanks for your work here @khl02007. 
- Were you able to declare all the tables in a vanilla mysql 8 docker? No key length issues? I would think you'd need to fetch from upstream to get the varchar reduction edits from #664 
- Did you try running `AnalysisNwbfile.cleanup()`? We had that secondary key issue with your table draft previously

khl02007 commented:
@CBroz1 
>     * Were you able to declare all the tables in a vanilla mysql 8 docker? No key length issues? I would think you'd need to fetch from upstream to get the varchar reduction edits from [Key length solve: Reduce varchars #664](https://github.com/LorenFrankLab/spyglass/pull/664)

I was. I'm not sure how recently I fetched from upstream but I will try this again just to make sure.

>     * Did you try running `AnalysisNwbfile.cleanup()`? We had that secondary key issue with your table draft previously

I haven not done this. I will try this.

samuelbray32 commented on src/spyglass/common/populate_all_common.py:
```diff
@@ -9,6 +9,7 @@
 from .common_nwbfile import Nwbfile
 from .common_session import Session
 from .common_task import TaskEpoch
+from spyglass.spikesorting.merge import SpikeSortingOutput
```
This import has issues.
Results in  a timeout when importing any part of spyglass since this file is imported as part of `spyglass.common.__init__.py`.  I'll dig in more and try to find where

khl02007 replied:
fixed

khl02007 commented:
@samuelbray32 

> **Higher level question** With the reduction of keys with `sorting_id` etc, what is the intended way of finding data through the merge table? In at least my analysis I rely on the `nwb_file_name`, and `sort_interval_name` (which in my code matches the corresponding `interval_list_name` for the epoch). Since those aren't restrictions that can be used in `SpikingOutput.fetch_nwb()` what is the correct way to find the data here?

Here is one way to do so:
```python
# starting from a merge_id
key = {"merge_id":"b4c89394-6f3b-49f6-d309-99e482571573"}

source = (SpikeSortingOutput & key).fetch1("source")
source_table = getattr(SpikeSortingOutput, source)
sorting_key = (source_table & key).fetch1()

# retrieve sorting as spikeinterface sorting
curated_sorting = CurationV1.get_sorting(sorting_key)
```

Do you think this is worth turning into a method? or include in tutorial notebook

samuelbray32 commented:
> @samuelbray32
> 
> > **Higher level question** With the reduction of keys with `sorting_id` etc, what is the intended way of finding data through the merge table? In at least my analysis I rely on the `nwb_file_name`, and `sort_interval_name` (which in my code matches the corresponding `interval_list_name` for the epoch). Since those aren't restrictions that can be used in `SpikingOutput.fetch_nwb()` what is the correct way to find the data here?
> 
> Here is one way to do so:
> 
> ```python
> # starting from a merge_id
> key = {"merge_id":"b4c89394-6f3b-49f6-d309-99e482571573"}
> 
> source = (SpikeSortingOutput & key).fetch1("source")
> source_table = getattr(SpikeSortingOutput, source)
> sorting_key = (source_table & key).fetch1()
> 
> # retrieve sorting as spikeinterface sorting
> curated_sorting = CurationV1.get_sorting(sorting_key)
> ```
> 
> Do you think this is worth turning into a method? or include in tutorial notebook

I meant more in the opposite direction, where I know the `nwb_file_name` and interval list, but want to connect it to a `merge_id`  or `sorting_id`.  

For example I can use `LFPOutput().fetch_nwb(restriction='nwb_file_name='Winnie20220714_.nwb"')` in the LFP pipeline to get data based on nwb_name.  

But that wouldn't work here since it isn't a key in `CurationV1`.

khl02007 commented:
> I meant more in the opposite direction, where I know the `nwb_file_name` and interval list, but want to connect it to a `merge_id` or `sorting_id`.
> 
> For example I can use `LFPOutput().fetch_nwb(restriction='nwb_file_name='Winnie20220714_.nwb"')` in the LFP pipeline to get data based on nwb_name.
> 
> But that wouldn't work here since it isn't a key in `CurationV1`.

Oh I see. The IDs are introduced at the level of Selection tables, so you would look it up there. For example, to get `sorting_id`,
```python
key = {
    "nwb_file_name":"wilbur20210326_.nwb",
    "interval_list_name":"wilbur20210326_.nwb_A_3070ce"
}
sorting_id = (SpikeSortingSelection & key).fetch1("sorting_id")
```

To get `merge_id`, you would also need a curation ID. Multiple curation IDs can be associated with a sorting ID, so you would have to know that in advance. Assuming you did, you would do
```python
key = {
    "nwb_file_name":"wilbur20210326_.nwb",
    "interval_list_name":"wilbur20210326_.nwb_A_3070ce",
    "curation_id" : 1
}

key["sorting_id"] = (SpikeSortingSelection & key).fetch1("sorting_id")
merge_id = (SpikeSortingOutput.CurationV1 & key).fetch1("merge_id")
```
And then you would call whichever function you need to get the data (`fetch_nwb`, `get_sorting` etc).

It is possible that you wouldn't know which child table of `SpikeSortingOutput` to run the query on in the last step. In that case I think we could probably do a join of the parent table with all the child tables, or iterate through each child table one by one. I thought `_Merge` has this feature but maybe I'm wrong?

khl02007 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,876 @@
+                    end_frame=interval_indices[1],
+                )
+                recordings_list.append(recording_single)
+                timestamps.append(
```
Good catch! fixed.

khl02007 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,876 @@
+
+        # slice in channels; include ref channel in first slice, then exclude it in second slice
+        if ref_channel_id >= 0:
+            recording = recording.channel_slice(channel_ids=channel_ids)
```
If `ref_channel` is not part of the sort group, then the recording obviously can't be referenced. So it seems like the error is expected. On the other hand, if you do correctly specify the `ref_channel` in your sort group, then you get back the non-ref channels after referencing (since the ref channel would just be zero). Is this not the correct behavior? What do you think it should be?

samuelbray32 replied:
I think this is an issue for something like tetrodes where you wank to reference the data on one tetrode against a reference on another.  Since it would be redundant to put the reference electrode in every group the previous version let you reference against one not in the group.  The `set_group_by_shank` function in the previous SortGroup table depended on this functionality

The previous version allowed it to work in the same way I suggested here. Previous version:
```
# include ref channel in first slice, then exclude it in second slice
if ref_channel_id >= 0:
    channel_ids_ref = np.append(channel_ids, ref_channel_id)
    recording = recording.channel_slice(channel_ids=channel_ids_ref)

    recording = si.preprocessing.common_reference(
        recording, reference="single", ref_channel_ids=ref_channel_id
    )
    recording = recording.channel_slice(channel_ids=channel_ids)
```

khl02007 replied:
I see, I made changes so that even if the ref channel is not one of the channels in the sort group, it is used to reference them.

khl02007 commented:
> @CBroz1
> 
> > ```
> > * Were you able to declare all the tables in a vanilla mysql 8 docker? No key length issues? I would think you'd need to fetch from upstream to get the varchar reduction edits from [Key length solve: Reduce varchars #664](https://github.com/LorenFrankLab/spyglass/pull/664)
> > ```
> 
> I was. I'm not sure how recently I fetched from upstream but I will try this again just to make sure.
> 
> > ```
> > * Did you try running `AnalysisNwbfile.cleanup()`? We had that secondary key issue with your table draft previously
> > ```
> 
> I haven not done this. I will try this.

khl02007 commented on src/spyglass/spikesorting/v1/curation.py:
```diff
@@ -0,0 +1,483 @@
+    """
+
+    @classmethod
+    def insert_curation(
```
I added a method called `insert_metric_curation` that does what you describe. I had to import the MetricCuration tables within the method to avoid circular import. Let me know if you think there are better solutions.

khl02007 commented on src/spyglass/spikesorting/merge.py:
```diff
@@ -0,0 +1,66 @@
+        key["merge_id"] = str(uuid.uuid4())
+        key["source"] = "ImportedSpikeSorting"
+        cls.insert1(key, ignore_extra_fields=True, skip_duplicates=True)
+        cls.ImportedSpikeSorting1.insert1(
```
Do you mean `insert_from_CurationV1`? In any case I also get same error. Curiously when I check `SpikeSortingOutput` and `SpikeSortingOutput.CuraitonV1`, the rows are populated. @samuelbray32 Do you know how to fix this? I don't know what's going on. Maybe this has to do with `_Merge`? perhaps @CBroz1 knows more.

edeno commented:
@CBroz1 could you look at:
1. the couple of questions on Merge table functioning
2. the general dependency structure of the tables (let's not worry about the specific functions within the tables)

CBroz1 commented on src/spyglass/spikesorting/v1/figurl_curation.py:
```diff
@@ -0,0 +1,267 @@
+                ),
+            ),
+            vv.MountainLayoutItem(
+                label="Spike amplitudes",
+                view=X.spike_amplitudes_view(
+                    unit_ids=X.unit_ids,
+                    _subsample_max_firing_rate=spike_amplitudes_subsample_max_firing_rate,
+                ),
+            ),
+            vv.MountainLayoutItem(
+                label="Autocorrelograms",
+                view=X.autocorrelograms_view(unit_ids=X.unit_ids),
+            ),
+            vv.MountainLayoutItem(
+                label="Cross correlograms",
+                view=X.cross_correlograms_view(unit_ids=X.unit_ids),
+            ),
+            vv.MountainLayoutItem(
+                label="Avg waveforms",
+                view=X.average_waveforms_view(unit_ids=X.unit_ids),
+            ),
+            vv.MountainLayoutItem(
+                label="Electrode geometry", view=X.electrode_geometry_view()
+            ),
+            # vv.MountainLayoutItem(
+            #    label='Unit similarity matrix',
+            #    view=unit_similarity_matrix_view
+            # ),
+            vv.MountainLayoutItem(
+                label="Curation", view=vv.SortingCuration2(), is_control=True
+            ),
+        ]
+    )
+    url_state = {
+        "initialSortingCuration": initial_curation_uri,
+        "sortingCuration": initial_curation_uri,
+    }
+    label = f"{recording_label} {sorting_label}"
+    url = view.url(label=label, state=url_state)
+    return url
```
```suggestion
    recording = R
    sorting = S

    sampling_frequency = recording.get_sampling_frequency()

    this_view = SpikeSortingView.create(
        recording=recording,
        sorting=sorting,
        segment_duration_sec=segment_duration_sec,
        snippet_len=(
            int(snippet_ms_before * sampling_frequency / 1000),
            int(snippet_ms_after * sampling_frequency / 1000),
        ),
        max_num_snippets_per_segment=max_num_snippets_per_segment,
        channel_neighborhood_size=channel_neighborhood_size,
    )

    # Assemble the views in a layout. Can be replaced with other layouts.
    raster_max_fire = raster_plot_subsample_max_firing_rate
    spike_amp_max_fire = spike_amplitudes_subsample_max_firing_rate

    sort_items = [
        vv.MountainLayoutItem(
            label="Summary", view=this_view.sorting_summary_view()
        ),
        vv.MountainLayoutItem(
            label="Units table",
            view=this_view.units_table_view(
                unit_ids=this_view.unit_ids, unit_metrics=unit_metrics
            ),
        ),
        vv.MountainLayoutItem(
            label="Raster plot",
            view=this_view.raster_plot_view(
                unit_ids=this_view.unit_ids,
                _subsample_max_firing_rate=raster_max_fire,
            ),
        ),
        vv.MountainLayoutItem(
            label="Spike amplitudes",
            view=this_view.spike_amplitudes_view(
                unit_ids=this_view.unit_ids,
                _subsample_max_firing_rate=spike_amp_max_fire,
            ),
        ),
        vv.MountainLayoutItem(
            label="Autocorrelograms",
            view=this_view.autocorrelograms_view(unit_ids=this_view.unit_ids),
        ),
        vv.MountainLayoutItem(
            label="Cross correlograms",
            view=this_view.cross_correlograms_view(unit_ids=this_view.unit_ids),
        ),
        vv.MountainLayoutItem(
            label="Avg waveforms",
            view=this_view.average_waveforms_view(unit_ids=this_view.unit_ids),
        ),
        vv.MountainLayoutItem(
            label="Electrode geometry",
            view=this_view.electrode_geometry_view(),
        ),
        vv.MountainLayoutItem(
            label="Curation", view=vv.SortingCuration2(), is_control=True
        ),
    ]

    return vv.MountainLayout(items=sort_items).url(
        label=f"{recording_label} {sorting_label}",
        state={
            "initialSortingCuration": initial_curation_uri,
            "sortingCuration": initial_curation_uri,
        },
    )
```

CBroz1 commented on src/spyglass/spikesorting/v1/figurl_curation.py:
```diff
@@ -0,0 +1,267 @@
+
+
+def _reformat_metrics(metrics: Dict[str, Dict[str, float]]) -> List[Dict]:
+    for metric_name in metrics:
+        metrics[metric_name] = {
+            str(unit_id): metric_value
+            for unit_id, metric_value in metrics[metric_name].items()
+        }
+    new_external_metrics = [
+        {
+            "name": metric_name,
+            "label": metric_name,
+            "tooltip": metric_name,
+            "data": metric,
+        }
+        for metric_name, metric in metrics.items()
+    ]
+    return new_external_metrics
```
```suggestion
    return [
        {
            "name": metric_name,
            "label": metric_name,
            "tooltip": metric_name,
            "data": {
                str(unit_id): metric_value
                for unit_id, metric_value in metric.items()
            },
        }
        for metric_name, metric in metrics.items()
    ]

```

CBroz1 commented on src/spyglass/spikesorting/v1/metric_curation.py:
```diff
@@ -0,0 +1,605 @@
+        key : dict
+            key for the inserted row
+        """
+        if len((cls & key).fetch()) > 0:
```
```suggestion
        if cls & key:
```

CBroz1 commented on src/spyglass/spikesorting/v1/metric_curation.py:
```diff
@@ -0,0 +1,605 @@
+import spikeinterface as si
+import spikeinterface.preprocessing as sp
+import spikeinterface.qualitymetrics as sq
+
```
```suggestion

from spyglass.settings import temp_dir
```

CBroz1 commented on src/spyglass/spikesorting/v1/metric_curation.py:
```diff
@@ -0,0 +1,605 @@
+    def make(self, key):
+        # FETCH
+        nwb_file_name = (
+            SpikeSortingSelection * MetricCurationSelection & key
+        ).fetch1("nwb_file_name")
+
+        waveform_params = (
+            WaveformParameters * MetricCurationSelection & key
+        ).fetch1("waveform_params")
+        metric_params = (
+            MetricParameters * MetricCurationSelection & key
+        ).fetch1("metric_params")
+        label_params, merge_params = (
+            MetricCurationParameters * MetricCurationSelection & key
+        ).fetch1("label_params", "merge_params")
+        sorting_id, curation_id = (MetricCurationSelection & key).fetch1(
+            "sorting_id", "curation_id"
+        )
+        # DO
+        # load recording and sorting
+        recording = CurationV1.get_recording(
+            {"sorting_id": sorting_id, "curation_id": curation_id}
+        )
+        sorting = CurationV1.get_sorting(
+            {"sorting_id": sorting_id, "curation_id": curation_id}
+        )
+        # extract waveforms
+        if "whiten" in waveform_params:
+            if waveform_params.pop("whiten"):
+                recording = sp.whiten(recording, dtype=np.float64)
+
+        waveforms_dir = (
+            os.environ.get("SPYGLASS_TEMP_DIR")
+            + "/"
+            + key["metric_curation_id"]
+        )
+        try:
+            os.mkdir(waveforms_dir)
+        except FileExistsError:
+            pass
```
the mkdir `-p` flag will ignore errors if already exists. `settings.py` already ensures that `temp_dir` exists

```suggestion
        (
            nwb_file_name,
            waveform_params,
            metric_params,
            label_params,
            merge_params,
            sorting_id,
            curation_id,
        ) = (
            SpikeSortingSelection
            * WaveformParameters
            * MetricParameters
            * MetricCurationParameters
            * MetricCurationSelection
            & key
        ).fetch1(
            "nwb_file_name",
            "waveform_params",
            "metric_params",
            "label_params",
            "merge_params",
            "sorting_id",
            "curation_id",
        )

        # DO
        # load recording and sorting
        curation_key = {"sorting_id": sorting_id, "curation_id": curation_id}
        recording = CurationV1.get_recording(curation_key)
        sorting = CurationV1.get_sorting(curation_key)
        # extract waveforms
        if waveform_params.get("whiten"):
            recording = sp.whiten(recording, dtype=np.float64)

        waveforms_dir = temp_dir + "/" + key["metric_curation_id"]
```

CBroz1 commented on src/spyglass/spikesorting/v1/metric_curation.py:
```diff
@@ -0,0 +1,605 @@
+        key : dict
+            primary key to MetricCuration
+        """
+        analysis_file_name = (cls & key).fetch1("analysis_file_name")
+        object_id = (cls & key).fetch1("object_id")
+        metric_param_name = (MetricCurationSelection & key).fetch1(
+            "metric_param_name"
+        )
+        metric_params = (
+            MetricParameters & {"metric_param_name": metric_param_name}
+        ).fetch1("metric_params")
+        metric_names = list(metric_params.keys())
+
+        metrics = {}
```
```suggestion
        analysis_file_name, object_id, metric_param_name, metric_params = (
            cls * MetricCurationSelection * MetricParameters & key
        ).fetch1(
            "analysis_file_name",
            "object_id",
            "metric_param_name",
            "metric_params",
        )
```

CBroz1 commented on src/spyglass/spikesorting/v1/metric_curation.py:
```diff
@@ -0,0 +1,605 @@
+        ) as io:
+            nwbf = io.read()
+            units = nwbf.objects[object_id].to_dataframe()
+        for metric_name in metric_names:
+            metrics[metric_name] = dict(zip(units.index, units[metric_name]))
+
+        return metrics
```
looping over a dict defaults to looping over keys
 ```suggestion
        return {
            name: dict(zip(units.index, units[name]))
            for name in metric_params
        }
```

CBroz1 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,878 @@
+    start, stop = start_indices[0], stop_indices[0]
+
+    # Loop through the rest of the intervals to join them if needed
+    for i in range(1, len(start_indices)):
+        next_start, next_stop = start_indices[i], stop_indices[i]
```
```suggestion
    for next_start, next_stop in zip(start_indices, stop_indices):
```

CBroz1 commented on src/spyglass/spikesorting/v1/sorting.py:
```diff
@@ -0,0 +1,353 @@
+        sorting_id : str
+            the unique sorting ID serving as primary key for SpikeSorting
+        """
+        if len((cls & key).fetch()) > 0:
```
see `query` approach above

CBroz1 commented on src/spyglass/spikesorting/v1/sorting.py:
```diff
@@ -0,0 +1,353 @@
+        # - information about the recording
+        # - artifact free intervals
+        # - spike sorter and sorter params
+        recording_key = (
```
see multi-fetch approach above

khl02007 commented on src/spyglass/spikesorting/v1/recording.py:
```diff
@@ -0,0 +1,876 @@
+    definition = """
+    # Set of electrodes to spike sort together
+    -> Session
+    sort_group_name: varchar(30)
```
OK I will just revert back to `sort_group_id:int` -- seems like we're having unneccessarily long discussions around this and I just want to keep @edeno happy at this point :)

khl02007 commented on src/spyglass/spikesorting/v1/figurl_curation.py:
```diff
@@ -0,0 +1,285 @@
+    definition = """
+    # Use `generate_curation_uri` method to generate a curation uri.
+    -> CurationV1
+    curation_uri: varchar(1000)     # GitHub-based URI to a file to which the manual curation will be saved
```
Introducing figurl curation id as discussed

edeno commented on src/spyglass/spikesorting/merge.py:
```diff
@@ -15,82 +15,55 @@
 class SpikeSortingOutput(_Merge):
     definition = """
-    # Output of spike sorting pipelines. Use `insert_from_TableName` method to insert rows.
+    # Output of spike sorting pipelines. Use `insert_from_sourcee` method to insert rows.
```
```suggestion
    # Output of spike sorting pipelines. Use `insert_from_source` method to insert rows.
```

CBroz1 commented on src/spyglass/spikesorting/v1/curation.py:
```diff
@@ -0,0 +1,504 @@
+    parent_curation_id=-1: int
+    -> AnalysisNwbfile
+    object_id: varchar(72)
+    merges_applied: enum("True", "False")
```
Sorry - this just occurred to me. It will accept python bools and represent them as 1/0

```suggestion
    merges_applied: bool
```

khl02007 commented:
Thanks @CBroz1 @edeno @samuelbray32 for the comments. I have incorporated as many suggested changes as I could, except for some that threw errors (I reverted those back) and those that were outdated and could not be applied (e.g. the insert sort group by shank function). But otherwise I think this is ready to be merged. The code can obviously be optimized further but I say let's go for it and worry about that later!
